### PR TITLE
Quick fix #79

### LINKED ISSUE #79: [bug | high priority] READY vs GUILD_CREATE race condition makes guild cache forever unavailable
### Description
Alchemy's event module is asynchronous, so two events received at different times can have Alchemy's internal handlers for them executed simultaneously. The guild cache is actually a group of `GenServer`s, exactly one for each guild, storing the state of its guild. A problem arises when you consider how the cache is started:
* `READY`'s handler spawns a GenServer for each guild cache, setting it to:
`%{"id" => "...", "unavailable" => true}`
* `GUILD_CREATE`'s handler checks if the GenServer for the relevant guild is running. If it is, then it replaces the cache with the incoming data. If it isn't, then it starts the GenServer with the new data.

Currently, it is very often the case that `READY` will finish starting all of the guilds' GenServers *after* `GUILD_CREATE` does, since they run simultaneously. This means that `GUILD_CREATE` causes the server to be started since it is not already, then filled with the guild's data. `READY` then completes its execution, starting the guilds' GenServers without checking if they are already running. This restarts the GenServers and sets them all as unavailable. There is no check in place to see if guilds are still unavailable after X seconds so it remains like this in guild cache forever, with updates to the cache through events just bouncing off of it since there is no structure to it.

#### Why not just change `READY`'s handler to only start the guild cache if it is not already started?

Consider this timeline of events:
```
1. READY is received and forwarded to event handlers async.
2. GUILD_CREATE is received and forwarded to event handlers async.
3. READY's event handler begins execution
4. GUILD_CREATE's event handler begins execution
5. READY's event handler checks if the GenServer is already running. It isn't
6. GUILD_CREATE's event handler starts the server with the new data, finishing execution
7. READY's event handler's check succeeded, so the GenServer is restarted as unavailable
8. The problem persists. The guild cache is now set as an unavailable placeholder forever
```

### Proposed Solutions

1. Assume that `READY` will always be received first (probably a fair assumption). Block the event dispatcher when `READY` is received by executing the handler synchronously. (Impl. [here](https://github.com/curz46/alchemy/commit/fd264edd1011dd041e243ab971b905f8962d5e70))  ([Credit](https://discordapp.com/channels/269508806759809042/269508806759809042/603280763357757443))
2. Rewrite the guild cache to work with a single `GenServer`, using one big map for all guilds. This has performance implications, because all operations on the guild cache are on a single process. Potentially significant when a bot is in a large number of guilds.
3. Start the guild cache GenServer only in response to a `GUILD_CREATE` event. Both guilds the bot is not in and guilds that are unavailable (whether that is at launch or otherwise) do not have GenServers running. Determining whether or not the bot user is in a guild can be done through a list of guild snowflakes managed independently of the guild cache.

Dylan

### PR DESCRIPTION
**Pull request by dcurgz**

### CODE REVIEW

File: lib/Discord/events.ex [hunk 1]
```diff
@@ -5,7 +5,7 @@ defmodule Alchemy.Discord.Events do
   # This module is then used by EventStage.Cache
```
Comment (cronokirby) [inline lib/Discord/events.ex:5]: Just remove the lines instead of commenting them out. We're using git so we don't have to comment things out :P

**Review by cronokirby — CHANGES_REQUESTED** [review]
The fix looks fine for now, but we can remove the dead code completely. We know where to look for in the git history to move it back later.

**Comment by jaspinegz** [comment]
👍 Removed

**Review by cronokirby — APPROVED** [review]
Nice

**Pull request merged**